gh-111964: Implement stop-the-world pauses

colesbury opened the pull request:
The `--disable-gil` builds occasionally need to pause all but one thread. Some examples include:

* Cyclic garbage collection, where this is often called a "stop the world event"
* Before calling `fork()`, to ensure a consistent state for internal data structures
* During interpreter shutdown, to ensure that daemon threads aren't accessing Python objects

This adds the following functions to implement global and per-interpreter pauses:

* `_PyRuntimeState_StopTheWorld` and `_PyRuntimeState_StartTheWorld`
* `_PyInterpreterState_StopTheWorld` and `_PyInterpreterState_StartTheWorld`

These functions are no-ops outside of the `--disable-gil` build.

This also adds `_PyRWMutex`, a "readers-writer" lock, which is used to serialize global stop-the-world pauses with per-interpreter pauses.

<!-- gh-issue-number: gh-111964 -->
* Issue: gh-111964
<!-- /gh-issue-number -->

colesbury committed:
gh-111964: Implement stop-the-world pauses

The `--disable-gil` builds occasionally need to pause all but one thread. Some
examples include:

* Cyclic garbage collection, where this is often called a "stop the world event"
* Before calling `fork()`, to ensure a consistent state for internal data structures
* During interpreter shutdown, to ensure that daemon threads aren't accessing Python objects

This adds the following functions to implement global and per-interpreter pauses:

* `_PyRuntimeState_StopTheWorld` and `_PyRuntimeState_StartTheWorld`
* `_PyInterpreterState_StopTheWorld` and `_PyInterpreterState_StartTheWorld`

These functions are no-ops outside of the `--disable-gil` build.

This also adds `_PyRWMutex`, a "readers-writer" lock, which is used to
serialize global stop-the-world pauses with per-interpreter pauses.

eendebakpt commented on Include/internal/pycore_interp.h:
```diff
@@ -39,6 +39,22 @@ struct _Py_long_state {
+struct _stoptheworld_state {
+    PyMutex mutex;       // Serializes stop-the-world attempts.
+
+    // NOTE: The below fields are protected by HEAD_LOCK(runtime), not by the
```
```suggestion
    // NOTE: The fields below are protected by HEAD_LOCK(runtime), not by the
```

eendebakpt commented on Include/internal/pycore_pystate.h:
```diff
@@ -138,13 +138,37 @@ _PyThreadState_GET(void)
+// "attached" state.
+extern void _PyThreadState_Park(PyThreadState *tstate);
+
+// Perform a stop-the-world pause for all threads in the all interpreters.
```
```suggestion
// Perform a stop-the-world pause for all threads in all interpreters.
```

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -196,6 +196,30 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+//
+// The design is optimized for simplicity of the implementation. The lock is
+// not fair: if fairness is desired, use an additional PyMutex to serialize
+// writers. The lock is also not reentrant.
```
Thanks for this extra note.  It's helpful.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -118,10 +118,15 @@ _PySemaphore_PlatformWait(_PySemaphore *sema, _PyTime_t timeout)
-
+        _PyTime_AsTimespec_clamp(deadline, &ts);
         err = sem_timedwait(&sema->platform_sem, &ts);
+ #endif
```
Is this change specific to the stop-the-world implementation?  It kind of seems like something worth doing regardless.

If so, something I do in that case sometimes is to put the worth-doing-on-its-own part in its own PR.  That way it shows up independently in `git log`.  (It also decouples the broader improvement from the specific one, which can factor in, for example, if the PR with the specific improvement gets reverted.  However, that usually doesn't matter all that much in practice.)

Having that distinct commit history entry might not be especially important most of the time, but occasionally it makes a difference for people tracking down the history of a feature (e.g. to understand it or to fix a bug) or browsing the commit history.  (I've been in that situation on a number of occasions.)  Making a separate PR is a bit more work (along with a little extra workflow annoyance), but the effort isn't enough to outweigh the benefit, IMHO.

I'm not saying you should necessarily split this out into a separate PR (assuming this *is* something that stands on its own), but it's something to think about.  If it were me, I'd do it.

ericsnowcurrently replied:
It would be a different story if we did straight merge commits instead of squash merges.  However, that's not what we do so making separate PRs is the solution.

colesbury replied:
This wasn't intentional included as part of this PR. I'll remove it.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -196,6 +196,30 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+
+// Write lock
+PyAPI_FUNC(void) _PyRWMutex_Lock(_PyRWMutex *rwmutex);
+PyAPI_FUNC(void) _PyRWMutex_Unlock(_PyRWMutex *rwmutex);
```
Similar to what I said about the change in parking_lot.c, consider adding `_PyRWMutex` in a distinct PR.

FWIW, *historically* we typically have avoided making a separate commit for an internal feature that we're adding strictly for the sake of a single other feature.  Usually the two are part of one commit, as you've done here.  However, I'm sure that's essentially a tradition rooted in the preference of early committers and the pain points of earlier workflows[^1] rather than deliberately decided for the sake of some technical/workflow benefit. [^2]

[^1]: Core dev workflow has changed significantly over the years and the 
[^2]: I think we've done it that way mostly because (1) people don't want to be bothered splitting up a branch into two (or more) and (2) it's easier to revert the whole.

Note that I've split up my own commits along these lines on many occasions in the past, without any controversy.  There are also plenty of times that I've kept the pieces together in a single commit, especially when the subsidiary internal piece has been tightly coupled to the actual change I'm aiming for.

Again, I'm not saying you necessarily need to split up this PR.  Just consider doing so.  Personally, I would.  At the end of the day it isn't critical.  From my perspective, the extra work of splitting up a branch/PR with distinct features is worth doing.

colesbury replied:
I'll split off `_PyRWMutex`.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -196,6 +196,30 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+//
+// The low two bits store whether the lock is write-locked (_Py_LOCKED) and
+// whether there are parked threads (_Py_HAS_PARKED). The remaining bits are
+// used to store the number of readers.
```
That implies a maximum of `sizeof(uintptr_t)/4` reader threads trying to hold the lock at once, right?  What if someone has more threads than that?  (just kidding 😄)  It probably isn't even physically possible to max out the counter.

colesbury replied:
Yeah, that's right.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,89 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+        // If the lock is not write-locked and there is no writer waiting, then
+        // we can increment the reader count.
+        if ((bits & (_Py_LOCKED|_Py_HAS_PARKED)) == 0) {
+            uintptr_t newval = bits + (1 << _PyRWMutex_READER_SHIFT);
```
It may seem silly and unnecessary, but consider spelling out the constraint here.  For example:

```suggestion
        if ((bits & (_Py_LOCKED|_Py_HAS_PARKED)) == 0) {
            assert((bits >> _PyRWMutex_READER_SHIFT) < (UINTPTR_MAX >> _PyRWMutex_READER_SHIFT));
            uintptr_t newval = bits + (1 << _PyRWMutex_READER_SHIFT);
```

colesbury replied:
I added an assertion.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -196,6 +196,30 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+// writers. The lock is also not reentrant.
+typedef struct {
+    uintptr_t bits;
+} _PyRWMutex;
```
If I've understood correctly, this is essentially a `Py_Mutex`, with the extra bits used for a counter.  Is that right?

A naive representation (for demonstration purposes):

```c
union {
    struct {
        uintptr_t _padding : UINTPTR_MAX-sizeof(PyMutex);
        PyMutex value;
    } mutex;
    struct {
#define _PyRWMUTEX_MUTEX_BITS 2
        uintptr_t value : sizeof(uintptr_t)-_PyRWMUTEX_MUTEX_BITS;
        uintptr_t _padding : _PyRWMUTEX_MUTEX_BITS;
    } num_readers;
    uintptr_t value;
}
```

<details>
<summary>or as accessors (not as clear):</summary>

```c
#define _PyRWMUTEX_AS_MUTEX(RWMUTEX) \
    ((PyMutex)RWMUTEX)

#define _PyRWMUTEX_MUTEX_BITS 2
#define _PY_RWMUTEX_GET_NUM_READERS(RWMUTEX_BITS) \
    (RWMUTEX_BITS >> _PyRWMUTEX_MUTEX_BITS)
```

</details>

Ideally, `_PyRWMUTEX_MUTEX_BITS` would be the bits needed for the largest bit flag used for `PyMutex`.  However, in this case we're using a subset of the possible `PyMutex` flags, so we match the largest flag in that subset.

Am I getting that right?  If so, it might help to be more clear in the doc comment that this is essentially a `PyMutex` with extra bits for the readers count and that we are commandeering some bits from the `PyMutex` for the count. 
 Having structural constraints in/near the definition here might help too (but perhaps would be hard to do clearly?).

colesbury replied:
> If I've understood correctly, this is essentially a Py_Mutex, with the extra bits used for a counter.

Yes, in the sense that both use two bits for state (locked and has_parked) and those bits are in the same positions. I think it's more helpful to think of it as "similar to a PyMutex" than embedding a PyMutex.

I would write the representation as:

```c
// e.g., 62 on 64-bit platforms
#define READER_COUNT_BITS (sizeof(uintptr_t)*CHAR_BIT - 2)

struct {
  bool is_write_locked: 1;
  bool has_parked: 1;
  uintptr_t reader_count: READER_COUNT_BITS;
};
````

> Ideally, _PyRWMUTEX_MUTEX_BITS would be the bits needed for the largest bit flag used for PyMutex. However, in this case we're using a subset of the possible PyMutex flags...

I'm not sure I understand what you mean. Ignoring the reader count for a moment, both `PyMutex` and `_PyRWMutex` only use two bits for state: is_locked? and has_parked?.

ericsnowcurrently replied:
I'll follow up on this on gh-112859.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,89 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+    for (;;) {
+        // If the lock is not write-locked and there is no writer waiting, then
+        // we can increment the reader count.
+        if ((bits & (_Py_LOCKED|_Py_HAS_PARKED)) == 0) {
```
(Please note that the comment I'm leaving here feels a big disorganized to me, but my intention is to try to nail down a feeling.  I'm definitely coming from a place of "I'm trying to make sense of this" and not "you've done something wrong".  I hope the tone reflects that.  When in doubt, assume curiosity rather than criticism. 😄)

When I was reading this code I found myself asking why we're using `_Py_LOCKED` and `_Py_HAS_PARKED` here.  Then I started thinking about how they effectively determine the value of `_PyRWMutex_READER_SHIFT` here even though they are (in my head) defined somewhere else and get treated as opaque here.  That also led me to puzzling out the relationship between `_PyRWMutex` and `PyMutex`.

Overall, something about the `_PyRWMutex` implementation feels off to me.  I'm going to walk through some general observations and then some specific ones.  I don't quite have my brain wrapped around all this to the point that I can provide clear and concise feedback, so I apologize if I cover unnecessary ground here.

First of all, the code doesn't feel as obvious as I would expect.  Instead, as a reader, I ended up hunting down the relevant context from unknown sources and piecing together the story as I went.  Perhaps that's due to how code is presented in a PR, where things might be more obvious as part of the whole file.  I'm not sure that's quite it though.

I also have some (mild?) concerns about the implicit coupling between `_PyRWMutex` and `PyMutex`, especially relative to fragility.  `_PyRWMutex` is clearly coupled to `PyMutex`, which makes sense to me now conceptually.  However, the coupling is quite implicit, with what seems like a fair degree of code duplication and some lack of shared structural elements that would help the compiler identify inconsistencies.  That presents maintenance risk, both relative to readers of the code and to resilience under any future external code changes (e.g. to `PyMutex`).  At the same time, I don't have a good sense of how much risk there actually might be.  My gut says there's enough that I ought to bring this up and that we should consider making some adjustments in the PR.

Here are some of the details I figured out:

* `_PyRWMutex_READER_SHIFT` is determined by the possible bit flag values we might store in the low bits; it must match the highest flag
* the set of flags may include RWMutex-specific flags (none currently) as well as ones defined externally
* `_PyRWMutex` is essentially a `PyMutex` with extra bits to track the count of readers, including some of the `PyMutex` bits that are repurposed for the count
* the bits that can be repurposed are basically `sizeof(PyMutex)*8 - _PyRWMutex_READER_SHIFT`
* we use a *subset* of the possible `PyMutex` bit flags, the highest of which affects `_PyRWMutex_READER_SHIFT`

Regarding the coupling between `_PyRWMutex` and `PyMutex`, the implementation isn't very explicit about it.  At first glance, it isn't clear that the two types are closely related, nor how.  Consequently, the use of `PyMutex` bit flags here feels somewhat untethered and it's tricky to reason about how those flags fit in and how they affect `_PyRWMutex_READER_SHIFT`.

Let me be a bit more concrete about coupling and potential fragility.  Changes to `PyMutex` (and its bit flags) could necessitate changes to the `PyRWMutex` implementation here that are easy to not notice and thus miss.  As things stand, the compiler wouldn't be able to help us out as much as it could.  Furthermore, it isn't obvious how the manual parts (e.g. the value of `_PyRWMutex_READER_SHIFT`, managing the use of the `PyMutex` flags, essentially duplicating parts of the `PyMutex` implementation) would need to be adjusted to the `PyMutex` changes.  Things might still compile but later crash, perhaps even only infrequently (in corner cases).

Here are two things to consider doing:

1. be more clear in the *code* about `PyMutex` (e.g. explicitly cast; embed `PyMutex` in `_PyRWMutex`)
2. be clear about the `PyMutex` bit flags we're using (we aren't using all possible flags); maybe don't get clever with repurposing unused bits from `PyMutex` (but might be unavoidable)

All that said, I recognize that some of this may have been done for performance reasons.  There's certainly an acceptable tradeoff involved here.  The question is, how much could we address coupling and context without impacting performance.  I expect there's room for that here, in part with code changes and in part with comments.

ericsnowcurrently replied:
FWIW, my initial question when I read this line was "What does `bits & (_Py_LOCKED|_Py_HAS_PARKED)` mean?"

colesbury replied:
Yeah, the `bits & (_Py_LOCKED|_Py_HAS_PARKED)` check is confusing. I've split it up  and added comments.

`_PyRWMutex` tries to follow some of the same conventions as other primitives in this file, but the only coupling is the re-use of macro constants. Otherwise, the implementations are independent. I've added `_Py_WRITE_LOCKED` and used that here to make the meaning of the state a bit more clear. 

We can't embed a `PyMutex` both because C is not rich enough to do it in a useful way, but also because `PyRWMutex` isn't a `PyMutex` + reader count. `PyMutex` supports a some things that we don't bother with here, like handoff.

ericsnowcurrently replied:
Sounds good.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,89 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+
+    if ((old_bits & _Py_HAS_PARKED) != 0) {
+        _PyParkingLot_UnparkAll(&rwmutex->bits);
+    }
```
I think I'm missing something here.  The assert (and the `0` we just set `bits` to) implies that we expect no waiting readers when `_PyRWMutex_Unlock()` is called.  That doesn't seem right.  Furthermore, if there are no readers then why would there be any parked threads to unpark?

colesbury replied:
The "reader count" only counts readers that hold the lock -- it doesn't include waiting readers. So if we (a writer) are unlocking the lock, there better not be any readers *also* holding the lock or we violated the internal constraint.

In other words, if `_Py_LOCKED` is set, then the reader count must be zero.

I'll add a comment in `lock.h` to make this more clear.

ericsnowcurrently commented on Include/pymacro.h:
```diff
@@ -160,6 +160,9 @@
     Py_FatalError("Unreachable C code path reached")
 #endif
 
+#define _Py_CONTAINER_OF(ptr, type, member) \
+    (type*)((char*)ptr - offsetof(type, member))
```
This is only valid for structs and arrays, right?  I thought we already had a macro like this somewhere.

colesbury replied:
Yeah, it's only valid for structs. The other one is in `pycore_llist.h`. I'll make that one use `_Py_CONTAINER_OF()` since they are exactly the same.

https://github.com/python/cpython/blob/16448cab44e23d350824e9ac75e699f5bcc48a14/Include/internal/pycore_llist.h#L40-L41

ericsnowcurrently commented on Include/internal/pycore_runtime.h:
```diff
@@ -229,6 +229,9 @@ typedef struct pyruntimestate {
     struct _faulthandler_runtime_state faulthandler;
     struct _tracemalloc_runtime_state tracemalloc;
 
+    _PyRWMutex stoptheworld_mutex;
+    struct _stoptheworld_state stoptheworld;
```
Why does this one have a mutex but the `PyInterpreterState` one doesn't?  (Note that I haven't looked at pystate.c yet.)

ericsnowcurrently replied:
Now I see.  It may be worth adding a comment indicating that the mutex is used by all interpreters, as well as global stop-the-world requests.

colesbury replied:
Yeah, that's right. I'll add a a comment.

ericsnowcurrently replied:
perfect

ericsnowcurrently commented on Include/internal/pycore_interp.h:
```diff
@@ -39,6 +39,22 @@ struct _Py_long_state {
+    Py_ssize_t thread_countdown;  // Number of threads that must pause.
+
+    PyThreadState *requester; // Thread that requested the pause (may be NULL).
+};
```
From what I can tell, stop-the-world is coupled with the eval breaker, so consider moving this to pycore_ceval_state.h with the other eval-related stuff.

colesbury replied:
It's only loosely related to the eval_breaker. It's fundamentally about `PyThreadState.state`: transitioning threads to the `_Py_THREAD_GC` state. You could still have this without eval-breaker (it would just be much less responsive), but you can't implement it without `PyThreadState.state`.

ericsnowcurrently replied:
Thanks for explaining.  It still feels like a ceval feature to me though.  Here's why:

* (as noted) it is executed via the eval breaker
* it has the same vibe as pretty much all the other things you find in Python/ceval_gil.c
* it is a replacement for functionality that the GIL provides currently (and the GIL is a ceval feature)

For a specific example, it feels kind of similar to `Py_AddPendingCall()`, which you'll find defined/implemented in ceval-related files.

As a counterpoint, the main motivation for stop-the-world is to facilitate GC.  GC is generally effected through the eval breaker, but GC isn't defined in ceval-related files; it *is* in its own feature-specific files.  (Stop-the-world doesn't feel like something that should have it's own files.)

Another counter example: stop-the-world definitely has a runtime lifecycle vibe too.  I'm reminded of `PyInterpreterState._finalizing` and `_PyRuntimeState._finalizing` (which both incidentally relate to a single thread state).  However, stop-the-world is all about thread states in an interpreter or the global runtime, rather than about those two state structs themselves, nor about their lifecycle.

> It's fundamentally about `PyThreadState.state`: transitioning threads to the _Py_THREAD_GC state.

Yeah, that makes sense.  Perhaps it should go in pycore_tstate.h.

Regardless, in the details this is certainly about more than transitioning `PyThreadState.state`.  It also includes actually pausing/unpausing threads, as well as establishing the state *all* threads should be in if any is created or tries to transition why the umbrella meta-state is in effect.

Finally, if I had to associate stop-the-world with a single feature, it would be code execution (specifically, synchronizing it, a la the GIL).  To me, "code execution" equates to ceval.

ericsnowcurrently replied:
FWIW, one thing that threw me off is that this would be the only case where we define a struct in pycore_interp.h that we also use in pycore_runtime.h (or vice versa).  There are certainly cases where we use use the same struct in both, but such structs are all defined in feature-specific header files.

ericsnowcurrently commented on Include/internal/pycore_interp.h:
```diff
@@ -164,6 +180,7 @@ struct _is {
 
     struct _warnings_runtime_state warnings;
     struct atexit_state atexit;
+    struct _stoptheworld_state stoptheworld;
```
Likewise, this would probably make more sense as a member of `struct _ceval_state` in pycore_ceval_state.h.

ericsnowcurrently replied:
\<retracted/\>

ericsnowcurrently commented on Include/internal/pycore_runtime.h:
```diff
@@ -229,6 +229,9 @@ typedef struct pyruntimestate {
     struct _tracemalloc_runtime_state tracemalloc;
 
+    _PyRWMutex stoptheworld_mutex;
+    struct _stoptheworld_state stoptheworld;
```
Likewise, this would probably make more sense as a member of `struct _ceval_runtime_state` in pycore_ceval_state.h.

ericsnowcurrently replied:
\<retracted/\>

ericsnowcurrently commented on Include/internal/pycore_pystate.h:
```diff
@@ -138,13 +138,37 @@ _PyThreadState_GET(void)
+//
+// NOTE: This is a no-op outside of Py_GIL_DISABLED builds.
+extern void _PyInterpreterState_StopTheWorld(PyInterpreterState *interp);
+extern void _PyInterpreterState_StartTheWorld(PyInterpreterState *interp);
```
The prefixes on these names don't seem quite right.  As noted, stop-the-world is tied to ceval more than to the interpreter/runtime itself.  Also, "stop the world" might not be as clear as some of the alternatives.

Here are some names to consider:

* change the prefix
   * `extern void _PyEval_StopTheWorldAll(_PyRuntimeState *runtime);`
   * `extern void _PyEval_StartTheWorldAll(_PyRuntimeState *runtime);`
   * `extern void _PyEval_StopTheWorld(PyInterpreterState *interp);`
   * `extern void _PyEval_StartTheWorld(PyInterpreterState *interp);`
* combine the pairs
   * `extern void _PyEval_StopTheWorld(PyInterpreterState *interp, int global);`
   * `extern void _PyEval_StartTheWorld(PyInterpreterState *interp, int global);`
* "StopTheWorld" -> "Pause"
   * `extern void _PyEval_PauseAll(_PyRuntimeState *runtime);`
   * `extern void _PyEval_UnpauseAll(_PyRuntimeState *runtime);`
   * `extern void _PyEval_PauseInterpreter(PyInterpreterState *interp);`
   * `extern void _PyEval_UnpauseInterpreter(PyInterpreterState *interp);`
* combined
   * `extern void _PyEval_Pause(PyInterpreterState *interp, int global);`
   * `extern void _PyEval_Unpause(PyInterpreterState *interp, int global);`

Regarding `Pause*`, an argument in its favor is that the name still works on the per-thread level (strictly hypothetically):

* `extern void _PyEval_PauseThread(PyThreadState *tstate);`
  (combined: `extern void _PyEval_Pause(PyThreadState *tstate, int level);`)
* `extern void _PyEval_UnpauseThread(PyThreadState *tstate);`
  (combined: `extern void _PyEval_UnpauseThread(PyThreadState *tstate, int level);`)

(FWIW, we'd probably use `_PyEval_PauseThread()` for something else.)

ericsnowcurrently replied:
Also, these would probably make more sense in pycore_ceval.h.

colesbury replied:
- I like the common prefix for the global and per-interpreter APIs.
- I don't think we should combine the APIs because you don't need a thread-state or interpreter state to start a global stop the world pause.

ericsnowcurrently replied:
My main reservation is that the `PyInterpreterState_` and `PyRuntimeState_` prefixes are used for either simple getters/setters or struct lifecycle.  For anything beyond that we either use a `Py_` prefix or a prefix that relates to a specific feature.  (We're a bit freer with the `PyThreadState_` prefix, but I'm okay with that.)  In this case I got the sense that this is a ceval feature, hence suggesting the `_PyEval_` prefix.

Regardless, let's find a prefix other than `_Py*State_`.

ericsnowcurrently replied:
FWIW, I'd still prefer names that are a bit more clear about what happens:

* "Suspend"
   * `_PyXXX_SuspendAllThreads(_PyRuntimeState *)`
   * `_PyXXX_ResumeAllThreads(_PyRuntimeState *)`
   * `_PyXXX_SuspendThreads(_PyRuntimeState *)`
   * `_PyXXX_ResumeThreads(_PyRuntimeState *)`
* combined (`NULL` means "all interpreters")
   * `_PyXXX_SuspendAllThreads(PyInterpreterState *)`
   * `_PyXXX_ResumeAllThreads(PyInterpreterState *)`

> you don't need a thread-state or interpreter state to start a global stop the world pause.

I don't doubt this, but I'm having trouble thinking of cases where I'm working without some "current" thread state.

What about:

 * `_PyThreadState_SuspendExcept(PyThreadState *tstate, int global)`
 * `_PyXXX_ResumeAllThreads(PyThreadState *tstate, int global)`

That's kind of like `_PyThreadState_DeleteExcept()`.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+#ifdef Py_GIL_DISABLED
+    start_the_world(&interp->stoptheworld);
+#endif
+}
```
Along the same lines as the header files, consider moving these (and the relevant static functions) to where we put all the other eval-breaker-driven code: Python/ceval_gil.c.

ericsnowcurrently replied:
Looking at it again, pystate.c seems just fine.

ericsnowcurrently commented on Include/internal/pycore_pystate.h:
```diff
@@ -138,13 +138,37 @@ _PyThreadState_GET(void)
+// "attached" state. It will switch to the "GC" state and pause until the
+// stop-the-world event completes, after which it will switch back to the
+// "attached" state.
+extern void _PyThreadState_Park(PyThreadState *tstate);
```
"Park" is a relatively unfamiliar term.  Consider this alternative:

```suggestion
extern void _PyEval_PauseThread(PyThreadState *tstate);
```

colesbury replied:
How about "suspend"? I think it has a bit more association with the idea that a different thread is in control.

I refactored the function to make it more clear that like `_PyThreadState_Detach()` and `_PyThreadState_Attach()` it corresponds to a state transition.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+    for (;;) {
+        // Switch threads that are detached to the GC stopped state
+        bool stopped_all_threads = park_detached_threads(stw);
+        HEAD_UNLOCK(runtime);
```
At this point new interpreters and thread states may be created, along with various objects.  Is that a problem?  Do we need to block in `new_threadstate()` and `_PyInterpreterState_New()`?

colesbury replied:
Yeah, dealing with new thread states is important (and missing, oops).

New interpreters aren't important themselves -- only the `PyThreadStates` they create.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+            }
+        }
+        else if (state == _Py_THREAD_ATTACHED && t != stw->requester) {
+            // TODO: set this per-thread, rather than per-interpreter.
```
Do we have a per-thread eval breaker?

colesbury replied:
Not yet -- that's why I left a TODO. I think my co-worker Brett may work on that.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+#define _Py_FOR_EACH_THREAD(stw)                                            \
+    for (PyInterpreterState *i = interp_for_stop_the_world((stw));          \
+            i != NULL; i = ((stw->is_global) ? i->next : NULL))             \
+        for (PyThreadState *t = i->threads.head; t; t = t->next)
```
I've always found the calling code to feel less magical if the variables are passed in:

```suggestion
#define _Py_FOR_EACH_THREAD(stw, i, t)                                            \
    for (i = interp_for_stop_the_world((stw));          \
            i != NULL; i = ((stw->is_global) ? i->next : NULL))             \
        for (t = i->threads.head; t; t = t->next)
```

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+    stw->thread_countdown = 0;
+    stw->requester = _PyThreadState_GET();  // may be NULL
+
+    _Py_FOR_EACH_THREAD(stw) {
```
```suggestion
    PyInterpreterState *i;
    PyThreadState *t;
    _Py_FOR_EACH_THREAD(stw, i, t) {
```

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+        }
+
+        int64_t wait_ns = 1000*1000;  // 1ms
+        if (PyEvent_WaitTimed(&stw->stop_event, wait_ns)) {
```
It might be worth adding a comment about why you picked 1ms.

colesbury replied:
I added a comment, but it might not inspire much confidence!

ericsnowcurrently replied:
better than nothing

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+        int64_t wait_ns = 1000*1000;  // 1ms
+        if (PyEvent_WaitTimed(&stw->stop_event, wait_ns)) {
+            assert(stw->thread_countdown == 0);
+            stw->stop_event = (PyEvent){0};
```
FWIW, it would be helpful if there were an indication of what setting the event to 0 means, whether a macro/function with a descriptive name or at least a comment.

colesbury replied:
I moved the initialization up and added a comment.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+    stw->world_stopped = 0;
+    stw->requester = NULL;
+    // Switch threads back to the detached state.
+    _Py_FOR_EACH_THREAD(stw) {
```
```suggestion
    PyInterpreterState *i;
    PyThreadState *t;
    _Py_FOR_EACH_THREAD(stw, i, t) {
```

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+        if (state == _Py_THREAD_GC &&
+            _Py_atomic_compare_exchange_int(&t->state,
+                                            &state,
+                                            _Py_THREAD_DETACHED)) {
```
If I understood correctly, all attached threads would have been detached by `stop_the_world()` and set to `_Py_THREAD_GC`, with is why we make this check and set this value.  Is that right?

ericsnowcurrently replied:
Does it matter that there might be some threads that aren't parked here (e.g. `_Py_THREAD_ATTACHED`)?  Presumably such threads would not have been included in GC operations, or for whatever else we might have stopped the world.

colesbury replied:
I'm not sure why I wrote it this way. All threads -- other than the requesting thread -- are in the `_Py_THREAD_GC` state. That should be an `assert(state == _Py_THREAD_GC)`, not an if-statement and we don't need a compare-exchange. I'll update this.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1922,21 +1942,238 @@ _PyThreadState_Attach(PyThreadState *tstate)
+    _PyRuntimeState *runtime = &_PyRuntime;
+
+    PyMutex_Lock(&stw->mutex);
+    if (stw->is_global) {
+        _PyRWMutex_Lock(&runtime->stoptheworld_mutex);
+    }
+    else {
+        _PyRWMutex_RLock(&runtime->stoptheworld_mutex);
+    }
```
Are we using `_PyRWMutex` here just so we can prioritize global stop-the-world requests over per-interpreter ones?  It doesn't actually have to do with read vs. write, does it?

colesbury replied:
Yeah, it doesn't have to do with "read vs. write". It's about preventing stop-the-world requests that overlap in the PyThreadStates they are trying to manage. Per-interpreter stop-the-world requests are independent of each other, so they can happen concurrently. Global stop-the-world requests affect all PyThreadStates, so they can't happen concurrently with other global stop-the-world requests nor with any per-interpreter stop-the-world requests.

C++ calls the same lock a [`shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex), with "shared" and "exclusive" locking. I think that name is less common, but the shared vs. exclusive terminology makes sense here:

- Global stop-the-world requests lock in "exclusive" mode - we can't have any ongoing per-interpreter ones ongoing.
- Per-interpreter stop-the-world-requests lock in "shared" mode - they can happen simultaneously in separate interpreters but prevent global stop-the-world requests.

The prioritization of global stop-the-world requests is "nice to have" but not fundamental.

ericsnowcurrently replied:
Thanks for explaining!

ericsnowcurrently reviewed (changes requested):
The change mostly makes sense, but it took me a little while to connect all the pieces in my head.  I've left various comments related to that and to a number of other things that might be worth adjusting.

colesbury committed:
Changes from review

colesbury reviewed (commented):
I split off the `_PyRWMutex` PR here: https://github.com/python/cpython/pull/112859. (This is now rebased on that one.)

I don't think the stop-the-world functionality belongs in `ceval` related files. It's fundamentally about managing `PyThreadState` objects (particularly, the "state" field) and is closely tied to the existing `_PyThreadState_Attach/Detach` functions. The connection to eval_breaker is only a small piece, similar to how the GC has an eval breaker bit.

colesbury committed:
Stop-the-world requests should block new threads.

Starting in the "suspended" state blocks any _PyThreadState_Attach()
call until the stop-the-world event completes.

ericsnowcurrently commented on Include/internal/pycore_pystate.h:
```diff
@@ -19,23 +19,27 @@ extern "C" {
+// [attached]       <->       [detached]       <->       [suspended]
+//   +                                                        ^
+//   +--------------------------------------------------------+
+//                          (bound thread)
```
FWIW, the text above makes it sound like thread a thread can go from "suspended" straight to "attached".

ericsnowcurrently replied:
Also, `(bound thread)` shows up at the top of the diagram and at the bottom.  I'm not sure how to interpret that.

colesbury replied:
By text above do you mean the diagram or the paragraph above it? I've updated both in an attempt to make them more clear.

The "(bound thread)" and "(stop-the-world thread)" labels indicate which thread can perform the transition. It's at the bottom because it's labeling the one-way transition from "attached" to "suspended".

ericsnowcurrently commented on Include/internal/pycore_pystate.h:
```diff
@@ -138,13 +142,37 @@ _PyThreadState_GET(void)
+// If there is no stop-the-world pause in progress, then this function is
+// a no-op. Returns one if the thread was switched to the "suspended" state and
+// zero otherwise.
+extern int _PyThreadState_Suspend(PyThreadState *tstate);
```
Would it make sense to fall back to detaching?

colesbury replied:
Yeah, I think that's conceptually simpler.

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1331,6 +1331,11 @@ init_threadstate(_PyThreadStateImpl *_tstate,
     tstate->datastack_limit = NULL;
     tstate->what_event = -1;
 
+    if (interp->stoptheworld.requested || _PyRuntime.stoptheworld.requested) {
+        // Start in the suspended state if there is an ongoing stop-the-world.
+        tstate->state = _Py_THREAD_SUSPENDED;
+    }
```
Should the thread be parked too (though probably in `_PyThreadState_Attach()` rather than here)?

colesbury replied:
That happens in `_PyThreadState_Attach()`. We don't want to block the thread here.

It is similar to how you can create a `PyThreadState` without holding the GIL, and only acquire the GIL (possibly blocking) when calling `PyEval_AcquireThread()` (or other similar functions that also call `_PyThreadState_Attach()` internally).

ericsnowcurrently reviewed (commented):
Mostly things are looking good.  There's the `_PyRWMutex` PR, I have one comment here, and there are a couple still open from my earlier review.

colesbury committed:
Update comments and _PyThreadState_Suspend.

_PyThreadState_Suspend now switches to "detached" if there is no active
stop-the-world request.

colesbury committed:
Rename stop-the-world functions

colesbury reviewed (commented):
I renamed the functions to `_PyEval_StopTheWorldAll`, etc. at least as a placeholder. I couldn't think of anything better. We could also go with only the `_Py` prefix.

ericsnowcurrently reviewed (approved):
Sorry for the delay.  LGTM

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -1830,21 +1855,237 @@ _PyThreadState_Attach(PyThreadState *tstate)
+}
+#endif  // Py_GIL_DISABLED
+
+void
+_PyEval_StopTheWorldAll(_PyRuntimeState *runtime)
+{
+#ifdef Py_GIL_DISABLED
+    stop_the_world(&runtime->stoptheworld);
+#endif
+}
+
+void
+_PyEval_StartTheWorldAll(_PyRuntimeState *runtime)
+{
+#ifdef Py_GIL_DISABLED
+    start_the_world(&runtime->stoptheworld);
+#endif
+}
+
+void
+_PyEval_StopTheWorld(PyInterpreterState *interp)
+{
+#ifdef Py_GIL_DISABLED
+    stop_the_world(&interp->stoptheworld);
+#endif
+}
+
+void
+_PyEval_StartTheWorld(PyInterpreterState *interp)
+{
+#ifdef Py_GIL_DISABLED
+    start_the_world(&interp->stoptheworld);
+#endif
+}
```
I still think these and the related static functions make more sense in Python/ceval_gil.c, but not enough to hold up this PR.

We can revisit some the discussion (including some of the alternate names I suggested) separately.

ericsnowcurrently reviewed (approved):
LGTM
